### PR TITLE
Implement function-references proposal and update GC proposal

### PR DESCRIPTION
**Pull request by eqrion**
Take 2 :)

### CODE REVIEW

File: crates/wast/src/resolve/names.rs [hunk 1]
```diff
@@ -1730,7 +1809,14 @@ impl<'a, 'b> ExprResolver<'a, 'b> {
             }
 
             LocalSet(i) | LocalGet(i) | LocalTee(i) => {
-                self.locals.resolve(i, "local")?;
+                for (depth, scope) in self.scopes.iter().enumerate().rev() {
+                    if let Err(e) = scope.resolve(i, "local") {
+                        if depth == 0 {
+                            return Err(e);
+                        }
+                    }
```
Comment (alexcrichton) [inline crates/wast/src/resolve/names.rs:1817]: Should this `break` once the local is resolved?

File: crates/wast/src/resolve/names.rs [hunk 2]
```diff
@@ -1328,32 +1326,35 @@ impl<'a> Module<'a> {
...
                     }
 
                     // .. followed by locals themselves
-                    for (id, _, _) in locals {
-                        resolver.locals.register(*id, "local")?;
+                    for local in locals {
+                        scope.register(local.id, "local")?;
                     }
 
+                    resolver.scopes.push(scope);
```
Comment (alexcrichton) [inline crates/wast/src/resolve/names.rs:1356]: Could you add a debug assert that after `resolve` below the length of `scopes` is 1?

File: crates/wast/src/resolve/names.rs [hunk 3]
```diff
@@ -1675,18 +1688,26 @@ impl<'a, T> Default for Namespace<'a, T> {
...
+    label: Option<Id<'a>>,
+    // Whether this block pushed a new scope for resolving locals
+    pushed_scope: bool,
+}
+
 struct ExprResolver<'a, 'b> {
     module: &'b Module<'a>,
-    locals: Namespace<'a, ()>,
-    labels: Vec<Option<Id<'a>>>,
+    scopes: Vec<Namespace<'a, ()>>,
```
Comment (alexcrichton) [inline crates/wast/src/resolve/names.rs:1701]: Mind adding a comment to this saying that the scopes here are for locals and that this is dynamically managed as we exit/enter `let` blocks?

File: crates/wast/src/resolve/names.rs [hunk 4]
```diff
@@ -1742,87 +1828,62 @@ impl<'a, 'b> ExprResolver<'a, 'b> {
...
                 }
 
-                // If the inline annotation persists to this point then resolve
-                // all of its inline value types.
-                if let Some(inline) = &mut bt.ty.inline {
-                    inline.resolve(self.module)?;
+                // Register all locals defined in this let
+                let mut scope = Namespace::default();
+                for local in &t.locals {
+                    scope.register(local.id, "local")?;
```
Comment (alexcrichton) [inline crates/wast/src/resolve/names.rs:1844]: The semantics of this implementation I think means that you can redefine a local name multiple times, it'll just shadow within a scope. Do you know if that's the intention of the `let` proposal? Or is it intended that shadowing is not allowed and redefining a name should be an error?

File: crates/wast/src/resolve/names.rs [hunk 5]
```diff
@@ -1730,7 +1809,14 @@ impl<'a, 'b> ExprResolver<'a, 'b> {
             }
 
             LocalSet(i) | LocalGet(i) | LocalTee(i) => {
-                self.locals.resolve(i, "local")?;
+                for (depth, scope) in self.scopes.iter().enumerate().rev() {
```
Comment (alexcrichton) [inline crates/wast/src/resolve/names.rs:1812]: Mind throwing an asert above this that `self.scopes.len() > 0`?

File: crates/wast/src/resolve/names.rs (hunk 1 above)
Comment (eqrion) [reply crates/wast/src/resolve/names.rs:1829]: Yup, nice catch. Impressed that it worked without that.

File: crates/wast/src/resolve/names.rs (hunk 4 above)
Comment (eqrion) [reply crates/wast/src/resolve/names.rs:1858]: Yes, that appears to be the intent of the proposal. I originally had it the other way around, and it failed a couple tests in `let.wast` which checked that shadowing was the behavior.

**Comment by eqrion** [comment]
@alexcrichton This should be ready to go now.

**Comment by alexcrichton** [comment]
Looks great to me, thanks again @eqrion!

Would y'all like a publish with this or did you want to reorganize the enum first?

**Comment by alexcrichton** [comment]
Also if you'd like, I'd be happy to add you as an owner of the crate on crates.io so you can push too (and push access here)

**Pull request merged**